src/ExceptionRenderer/RendererAbstract.php

```diff
@@ -0,0 +1,139 @@
+        $this->processStackTrace();
+        $this->processPreviousException();
+    }
+
+    public function __toString(): string
+    {
+        try {
+            $this->processAll();
+
+            return $this->output;
+        } catch (\Throwable $e) {
+            // fallback if Exception occur in renderer
+            return get_class($this->exception).' ['.$this->exception->getCode().'] Error:'.$this->_($this->exception->getMessage());
+        }
+    }
+
+    protected function replaceTokens(array $tokens, string $text): string
+    {
+        return str_replace(array_keys($tokens), array_values($tokens), $text);
+    }
+
+    protected function parseCallTraceObject($call): array
+    {
+        $parsed = [
+            'line'             => (string) ($call['line'] ?? ''),
+            'file'             => (string) ($call['file'] ?? ''),
+            'class'            => $call['class'] ?? null,
+            'object'           => $call['object'] ?? null,
+            'function'         => $call['function'] ?? null,
+            'args'             => $call['args'] ?? [],
+            'object_formatted' => null,
+            'file_formatted'   => null,
+            'line_formatted'   => null,
+        ];
+
+        $parsed['file_formatted'] = str_pad(substr($parsed['file'], -40), 40, ' ', STR_PAD_LEFT);
+        $parsed['line_formatted'] = str_pad($parsed['line'] ?? '', 4, ' ', STR_PAD_LEFT);
+
+        if (null !== $parsed['object']) {
+            $parsed['object_formatted'] = $parsed['object']->name ?? get_class($parsed['object']);
+        }
+
+        return $parsed;
+    }
+
+    public static function toSafeString($val): string
+    {
+        if (is_object($val) && !$val instanceof \Closure) {
+            return isset($val->_trackableTrait)
+                ? get_class($val).' ('.$val->name.')'
+                : 'Object '.get_class($val);
+        }
+
+        return (string) json_encode($val);
+    }
+
+    protected static function getClassShortName(\Throwable $exception): string
+    {
+        return preg_replace('/.*\\\\/', '', get_class($exception));
+    }
+
+    /**
+     * @return string
+     */
+    protected function getExceptionTitle(): string
+    {
+        return $this->is_atk_exception
+            ? $this->exception->getCustomExceptionTitle()
+            : static::getClassShortName($this->exception).' Error';
+    }
+
+    /**
+     * @return string
+     */
+    protected function getExceptionName(): string
+    {
+        return $this->is_atk_exception
+            ? $this->exception->getCustomExceptionName()
+            : get_class($this->exception);
+    }
+
+    public function _($message, array $parameters = [], ?string $domain = null, ?string $locale = null): string
+    {
+        return $this->adapter
+            ? $this->adapter->_($message, $parameters, $domain, $locale)
+            : Translator::instance()->_($message, $parameters, $domain, $locale);
+    }
+}
```

Comment:
If a `getAdapter` or `getTranslationAdapter` method would be added to the exception class this could be basically removed.
Mock implementation in the Exception class:
```php
public function getTranslationAdapter(): ITranslatorAdapter {
    return $this->adapter ?? Translator::instance();
}
```
which could be later used here like so:
```php
if ($this->exception instanceof \atk4\core\Exception) // Maybe make it into an interface instead for use in exceptions, or a trait
    return $this->exception->getTranslationAdapter()->_($message, $parameters, $domain, $locale);

// We don't wanna translate other exceptions
return $message;
```

Reply:
`Translator::instance()` is not an `ITranslatorAdapter`, it can be because is conform to Interface but is not an `ITranslatorAdapter`, because is inteded to be used with the setted `ITranslatorAdapter`

Why added something that is intended to stay there forever after set?

`Translator -> setAdapter ->_($translate)`
or
`new Adapter -> configure adapter...`
`Translator -> set adapter (already configured) -> _($translate)`

Probably i'm too defensive, but IMHO if you put getTranslator you leave too much space for possible strange and wrong usage of the class